### PR TITLE
[KED-2381] Improve responsive zoom approach

### PR DESCRIPTION
**Pull request by bru5**
## Description

Improves responsive zoom approach to work as follows, in order of precedence

- once, after any graph or sidebar change
  - fit whole graph
  - centre whole graph
  - crop graph when reaching a minimum usable scale
- once, after an element is selected
  - respect the above constraints, and
  - ensure node is clearly visible between panels
  - apply an increased minimum scale
- once, after a browser resize
  - respect all the above constraints
- at any other time
  - allow user pan and zoom freely within extents

## Development notes

This functionality is a usability improvement and is much needed as new sidebars are being added.

A follow up PR will refactor the view system in general now that it is outgrowing its current shape.

## QA notes

Use the above description as a guide for requirements and testing.

Note that some edge cases may not appear to be an 'optimal' view, but none should be broken.

This will include testing in various combinations:

- open / closed sidebar and metadata panel
- resizing the browser to various viewport sizes / aspect ratios
- different graph aspect ratios (square, wide, tall)
- selecting elements on the sidebar and on the chart
- selecting tags and pipelines
- toggling labels

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

## Legal notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

**Review by richardwestenra — APPROVED** [review]
The only comment I'd have made is to refactor things a little to break the big blocks of code into new functions/methods but you're gonna do that in a separate PR 👍

File: src/selectors/layout.test.js [hunk 1]
```diff
@@ -25,29 +25,12 @@ describe('Selectors', () => {
...
-        expect(getSidebarWidth(true, 1200, sidebarWidth)).toEqual(open);
-        expect(getSidebarWidth(true, 900, sidebarWidth)).toEqual(open);
-      });
-
-      it(`sets sidebar width to ${closed} on screens smaller than ${sidebarBreakpoint}`, () => {
-        expect(getSidebarWidth(true, 480, sidebarWidth)).toEqual(closed);
-        expect(getSidebarWidth(true, 320, sidebarWidth)).toEqual(closed);
-      });
+    it(`if visible is true returns the 'open' width`, () => {
+      expect(getSidebarWidth(true, sidebarWidth)).toEqual(sidebarWidth.open);
```
Comment (studioswong) [inline src/selectors/layout.test.js:29]: I'm not entirely sure if the new test covers the same scenerios as covered by the old tests with the different break points - the tests seems more like a generic unit test?

**Comment by studioswong** [comment]
Thanks @bru5 , the new improvement looks good! I added a comment about the new test setup as it seems to be too generic and doesn't really cover the scenerios that this PR has set out to solve, but I guess that could be covered in your next PR if it also touches on the same sidebar concerns.

On another note, when you mentioned that new sidebars are added, are you referring to the new 'panels' on the right ( such as code panels, etc) are added?

**Review by studioswong — APPROVED** [review]

File: src/selectors/layout.test.js (hunk 1 above)
Comment (bru5) [reply src/selectors/layout.test.js:29]: The function changed so the old tests here were no longer relevant, but good catch anyway.

**Pull request merged**